numpy/core/src/umath/reduction.c

```diff
@@ -313,12 +325,40 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
      * Initialize the result to the reduction unit if possible,
      * otherwise copy the initial values and get a view to the rest.
      */
-    if (identity != Py_None) {
-        if (PyArray_FillWithScalar(result, identity) < 0) {
+    if (initial != NULL && initial != Py_None) {
+        /*
+         * User provided an `initial` value and it is not `None`.
+         * NOTE: It may make sense to accept array-valued `initial`,
```

Comment:
Note the suggestion of an array-valued `initial` for `bincount` at https://github.com/numpy/numpy/issues/22471#issuecomment-1288176870

Reply:
Yeah, but it isn't important for this PR (but I could prioritize it for a follow-up!).  I am planning on rewording things to use `initial`, and then say whether it is "identity" or "default".  And also allow the flexibility of using identity (only when reduction is non-empty) rather than default (only when it is empty).
The example being `-inf` for `max()`, when we want `max([])` to error (although that is likely not what we want!).

Just busy rewording, the main thing here (from my point of view) is still whether "reorderable" should just be considered "static" (mainly to simplify the `get_identity()`/`get_reduction_initial()` function), although the dynamic makes it a bit easier to fall back to the ufunc!

Reply:
OK, sounds good. It does seem to me that `reorderable` can be static for any given loop -- looking at this briefly again, the combination with the identify continues to feel odd.

Reply:
Thanks.  OK, I got some more refactoring to do here, then will see if there is a good way to disentangle it and make the info static.
I do think it is fathomable that it isn't static (i.e. dtype instance dependent), but the main thing will be if it seems too clunky w.r.t. to falling back to the current mechanism.